File: Poj/1151.pas
Const
  Fin = '';
  Fout = '';

Type
  Link = ^Obj;
  Obj = Record
    a, b: Longint;
    l, r: Link;
    c: Longint;
    s: Real;
End;

Type
  Line = Record
    a, b, c: Real;
    d: Boolean;
End;

Var
  Top: Link;
  i: Longint;
  ans: Real;
  x1, x: Array[1..400]Of Real;
  l: Array[1..200]Of Line;
  n: Longint;
  j, k, p, q: Real;
  cnt :Longint;
  ii: Longint;
Procedure Qsort(p, r: Longint);
Var
  xx: Real;
  i, j: Longint;
Begin
  xx := x1[p];
  i := p;
  j := r;
  While (i < j) Do
  Begin
    While (i < j) And (x1[j] > xx) Do Dec(j);
    If (i < j) Then Begin x1[i] := x1[j]; Inc(i); End;
    While (i < j) And (x1[i] < xx) Do Inc(i);
    If (i < j) Then Begin x1[j] := x1[i]; Dec(j); End;
  End;
  x1[i] := xx;
  If p < i - 1 Then Qsort(p, i - 1);
  If i + 1 < r Then Qsort(i + 1, r);
End;

Function Find(q: Real): Longint;
Var
  i, j: Longint;
  m: Longint;
Begin
  i := 1;
  j := cnt;
  While True Do
  Begin
    m := (i + j) div 2;
    If x[m] = q Then Begin Find := m; Exit; End;
    If q < x[m] Then j := m - 1
    Else i := m + 1;
  End;
End;

Procedure Qsortl(p, r: Longint);
Var
  x: Line;
  i, j: Longint;
Begin
  x := l[p];
  i := p;
  j := r;
  While i < j Do
  Begin
    While (i < j) And (l[j].c > x.c) Do Dec(j);
    If i < j Then Begin l[i] := l[j]; Inc(i); End;
    While (i < j) And (l[i].c < x.c) Do Inc(i);
    If i < j Then Begin l[j] := l[i]; Dec(j); End;
  End;
  l[i] := x;
  If p < i - 1 Then Qsortl(p, i - 1);
  If i + 1 < r Then Qsortl(i + 1, r);
End;

Procedure Make(Var Tree: Link; a, b: Longint);
Var
  m: Longint;
Begin
  New(Tree);
  Tree^.a := a;
  Tree^.b := b;
  Tree^.c := 0;
  Tree^.s := 0;
  If a < b Then
  Begin
    m := (a + b) div 2;
    Make(Tree^.l, a, m);
    Make(Tree^.r, m + 1, b);
  End;
End;

Procedure Updata(Tree: Link);
Begin
  If Tree^.c > 0 Then Tree^.s := x[Tree^.b + 1] - x[Tree^.a]
  Else If (Tree^.a = Tree^.b) Then Tree^.s := 0
  Else Tree^.s := Tree^.l^.s + Tree^.r^.s;
End;

Procedure Insert(Tree: Link; a, b: Longint);
Var
  m: Longint;
Begin
  If (a <= Tree^.a) And (Tree^.b <= b) Then
  Begin
    Inc(Tree^.c);
    Updata(Tree);
  End
  Else
  Begin
    m := (Tree^.a + Tree^.b) div 2;
    If a <= m then Insert(Tree^.l, a, b);
    If b > m Then Insert(Tree^.r, a, b);
    Updata(Tree);
  End;
End;

Procedure Delete(Tree: Link; a, b: Longint);
Var
  m: Longint;
Begin
  If (a <= Tree^.a) And (Tree^.b <= b) Then
  Begin
    Dec(Tree^.c);
    Updata(Tree);
  End
  Else
  Begin
    m := (Tree^.a + Tree^.b) div 2;
    If a <= m Then Delete(Tree^.l, a, b);
    If b > m Then Delete(Tree^.r, a, b);
    Updata(Tree);
  End;
End;

Begin
  Assign(Input, Fin);
  Assign(Output, Fout);
  Reset(Input);
  Rewrite(Output);

  ii := 0;
  While True Do
  Begin
    Read(n);
    If n = 0 Then
    Begin
      Close(Output);
      Exit;
    End;
    For i:=1 To n Do
    Begin
      Read(j, k, p, q);
      l[i].a := j;
      l[i].b := p;
      l[i].c := k;
      l[i].d := True;

      l[i + n].a := j;
      l[i + n].b := p;
      l[i + n].c := q;
      l[i + n].d := False;

      x1[i] := j;
      x1[i + n] := p;
    End;

    Qsort(1, 2 * n);
    x[1] := x1[1];
    cnt := 1;
    For i:=2 To 2 * n Do
      If x1[i] <> x1[i - 1] Then
      Begin
        Inc(cnt);
        x[cnt] := x1[i];
      End;
    For i:=1 To 2 * n Do
    Begin
      l[i].a := Find(l[i].a);
      l[i].b := Find(l[i].b);
      l[i].b := l[i].b - 1;
    End;

    Qsortl(1, 2 * n);
    Make(Top, 1, cnt - 1);
    ans := 0;
    For i:=1 To 2 * n - 1Do
    Begin
      If l[i].d Then Insert(Top, Trunc(l[i].a), Round(l[i].b))
        Else Delete(Top, Round(l[i].a), Round(l[i].b));
      ans :=  ans + (Top^.s) * (l[i + 1].c - l[i].c);
    End;

    Inc(ii);
    WriteLn('Test case #', ii);
    Write('Total explored area: ');
    WriteLn(ans: 0: 2);
    WriteLn;
  End;
End.

File: Poj/1151.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <algorithm>
#define rep(i, j) for (int i = 0; i < j; i++)

using namespace std;

struct obj
{
    int a, b;
    obj *l, *r;
    double s;
    int add;
}f[11111];

struct obj1
{
    double a, b;
    double c;
    bool flg;
}l[11111];

int cnt;
double x[1111];
double tmp[1111];
obj* top;
int cnt1;

obj* make(int a, int b)
{
    cnt++;
    int tmp = cnt;
    f[cnt].a = a;
    f[cnt].b = b;
    f[cnt].l = NULL;
    f[cnt].r = NULL;
    f[cnt].s = 0;
    f[cnt].add = 0;
    if (a < b)
    {
        int mid = (a + b) >> 1;
        f[tmp].l = make(a, mid);
        f[tmp].r = make(mid + 1, b);
    }
    return f + tmp;
}

void update(obj* a)
{
    if (a->add > 0) a->s = x[a->b + 1] - x[a->a];
    else if (a->a == a->b) a->s = 0;
    else a->s = a->l->s + a->r->s;
}

void insert(obj* t, int a, int b)
{
    if (a <= t->a && t->b <= b)
    {
        (t->add)++;
        update(t);
    }
    else
    {
        int mid = (t->a + t->b) >> 1;
        if (a <= mid) insert(t->l, a, b);
        if (b > mid) insert(t->r, a, b);
        update(t);
    }
}

void del(obj* t, int a, int b)
{
    if (a <= t->a && t->b <= b)
    {
        (t->add)--;
        update(t);
    }
    else
    {
        int mid = (t->a + t->b) >> 1;
        if (a <= mid) del(t->l, a, b);
        if (b > mid) del(t->r, a, b);
        update(t);
    }
}

int cmp(obj1 a, obj1 b)
{
    return a.c < b.c;
}


int find(double xx)
{
    int l = 0, r = cnt1 - 1;
    while(l < r - 1)
    {
        int mid = (l + r) >> 1;
        if (xx < x[mid]) r = mid; else l = mid;
    }
    if (x[l] == xx) return l;
    else return r;
}
void test(obj* t)
{
    printf("%d %d\n", t->a, t->b);
    if (t->a == t->b) return;
    test(t->l);
    test(t->r);
}
int main()
{
    int n;
    int ii = 1;
    while(true)
    {
        scanf("%d", &n);
        if (n == 0) break;
        double a, b, c, d;
        rep(i, n)
        {
            scanf("%lf%lf%lf%lf", &a, &b, &c, &d);
            l[i].a = a;
            l[i].b = c;
            l[i].c = b;
            l[i].flg = true;
            
            l[i + n].a = a;
            l[i + n].b = c;
            l[i + n].c = d;
            l[i + n].flg = false;
            
            x[i] = a;
            x[i + n] = c;
        }
        sort(x, x + 2 * n);
        cnt = 0;
        cnt1 = 0;
        rep(i, 2 * n) if (i == 0 || x[i] != x[i - 1]) tmp[cnt1++] = x[i];
        rep(i, cnt1) x[i] = tmp[i];
        sort(l, l + 2 * n, cmp);
        double ans = 0;
        top = make(0, cnt1 - 1);
        rep(i, 2 * n)
        {
            if(l[i].flg) insert(top, find(l[i].a), find(l[i].b) - 1);
            else del(top, find(l[i].a), find(l[i].b) - 1);
            ans += top->s * (l[i + 1].c - l[i].c);
        }
        printf("Test case #%d\n", ii++);
        printf("Total explored area: %.2f\n", ans);
        putchar('\n');
    }
}